SDWebImage/SDWebImageCodersManager.m

```diff
@@ -34,48 +36,49 @@ + (nonnull instancetype)sharedInstance {
 - (instancetype)init {
     if (self = [super init]) {
         // initialize with default coders
-        _mutableCoders = [@[[SDWebImageImageIOCoder sharedCoder]] mutableCopy];
+        NSMutableArray<id<SDWebImageCoder>> *mutableCoders = [@[[SDWebImageImageIOCoder sharedCoder]] mutableCopy];
 #ifdef SD_WEBP
-        [_mutableCoders addObject:[SDWebImageWebPCoder sharedCoder]];
+        [mutableCoders addObject:[SDWebImageWebPCoder sharedCoder]];
 #endif
-        _mutableCodersAccessQueue = dispatch_queue_create("com.hackemist.SDWebImageCodersManager", DISPATCH_QUEUE_CONCURRENT);
+        _coders = [mutableCoders copy];
+        _codersLock = dispatch_semaphore_create(1);
```

Comment:
Just a theoretical question: are we sure we don't need to synchronise the creation of the array? I know in theory you only get the `SDWebImageCodersManager` instance after `init`, but I'm not sure what a concurrent environment can cause. Just going super-theoretical here :) The original implementation was the same. But the bullet proof one is if you have a non-thread-safe resource, synchronise all the access to that resource, including creation and deallocation.
-- No need to change anything --

Reply:
`init` method in Objective-C is always safe. Because you create object use 2-way-initilizer (First call `alloc`, then `init`). If you call `init` twice, you get 2 different instances.
Thread-safe is about different threads modify the same instance(or variable), but not different threads modify different instances..

Reply:
But you should note it's not safe if you initilizer (Or something like `Class convenience initilizer`) which share a global static variable. At this time you should take responsibility to access that shared variable thread-safe.

But most of our initilizer does not use that. Only rare class use shared instance (Like `- [SDAnimatedImage imageNamed:]`, which use semaphore to guard. Or something like `defaultDownloadConfig`, which use `dispatch_once` to guard).

Note this hard is not useful in Swift. Swift initilizer or class static variable is always thread-safe. (Swift compiler keep this variable thread-safe by lock internally)

Reply:
I know :) I meant a situation where `alloc` was executed, so we have an instance and on the same instance `init` on one thread and another method on another thread go concurrent. But that is impossible